chirho/robust/internals/linearize.py

```diff
@@ -0,0 +1,158 @@
+        #     break
+    return x
+
+
+def conjugate_gradient_solve(f_Ax: Callable[[T], T], b: T, **kwargs) -> T:
+    flatten, unflatten = make_flatten_unflatten(b)
+
+    def f_Ax_flat(v: torch.Tensor) -> torch.Tensor:
+        v_unflattened: T = unflatten(v)
+        result_unflattened = f_Ax(v_unflattened)
+        return flatten(result_unflattened)
+
+    return unflatten(_flat_conjugate_gradient_solve(f_Ax_flat, flatten(b), **kwargs))
+
+
+def make_empirical_fisher_vp(
+    func_log_prob: Callable[Concatenate[ParamDict, Point[T], P], torch.Tensor],
+    log_prob_params: ParamDict,
+    data: Point[T],
+    *args: P.args,
+    **kwargs: P.kwargs,
+) -> Callable[[ParamDict], ParamDict]:
+    batched_func_log_prob: Callable[[ParamDict, Point[T]], torch.Tensor] = torch.vmap(
+        lambda p, data: func_log_prob(p, data, *args, **kwargs),
+        in_dims=(None, 0),
+        randomness="different",
+    )
+
+    def bound_batched_func_log_prob(params: ParamDict) -> torch.Tensor:
+        return batched_func_log_prob(params, data)
+
+    def jvp_fn(v: ParamDict) -> torch.Tensor:
+        return torch.func.jvp(bound_batched_func_log_prob, (log_prob_params,), (v,))[1]
+
+    vjp_fn = torch.func.vjp(bound_batched_func_log_prob, log_prob_params)[1]
+
+    def _empirical_fisher_vp(v: ParamDict) -> ParamDict:
+        jvp_log_prob_v = jvp_fn(v)
+        return vjp_fn(jvp_log_prob_v / jvp_log_prob_v.shape[0])[0]
+
+    return _empirical_fisher_vp
+
+
+def linearize(
+    model: Callable[P, Any],
+    guide: Callable[P, Any],
+    *,
+    num_samples_outer: int,
+    num_samples_inner: Optional[int] = None,
+    max_plate_nesting: Optional[int] = None,
+    cg_iters: Optional[int] = None,
+    residual_tol: float = 1e-10,
+) -> Callable[Concatenate[Point[T], P], ParamDict]:
+    assert isinstance(model, torch.nn.Module)
+    assert isinstance(guide, torch.nn.Module)
+    if num_samples_inner is None:
+        num_samples_inner = num_samples_outer**2
+
+    predictive = pyro.infer.Predictive(
+        model,
+        guide=guide,
+        num_samples=num_samples_outer,
+        parallel=True,
+    )
+    predictive_params, func_predictive = make_functional_call(predictive)
+
+    log_prob = NMCLogPredictiveLikelihood(
+        model, guide, num_samples=num_samples_inner, max_plate_nesting=max_plate_nesting
+    )
+    log_prob_params, func_log_prob = make_functional_call(log_prob)
+    score_fn = torch.func.grad(func_log_prob)
+
+    cg_solver = functools.partial(
+        conjugate_gradient_solve, cg_iters=cg_iters, residual_tol=residual_tol
+    )
+
+    @functools.wraps(score_fn)
+    def _fn(point: Point[T], *args: P.args, **kwargs: P.kwargs) -> ParamDict:
+        with torch.no_grad():
+            data: Point[T] = func_predictive(predictive_params, *args, **kwargs)
+            data = {k: data[k] for k in point.keys()}
+        fvp = make_empirical_fisher_vp(
+            func_log_prob, log_prob_params, data, *args, **kwargs
+        )
```

Comment:
I think these lines should be pulled out of `_fn` because (1) each time `_fn` is called a different approximation to the fisher information matrix is used since `data` is resampled, and (2) there's a performance hit because you need to resample `data` each time. This will be time consuming when `data` has a lot of Monte Carlo samples.

Reply:
I don't think we can move these lines out of `_fn` because they depend on the values of `args` and `kwargs`, but I think we'd want a fresh approximation to the Fisher matrix every time anyway since `log_prob_params` could change (e.g. by a gradient update) even if `args` didn't.

Reply:
If we put *args and **kwargs as inputs to linearize, then we could pull out these lines, right? I think the odd behavior is that the resulting influence function is not consistent across points. For example, if I call `_fn` from linearize on the same point twice, you'll get two different answers. Similarly, it's harder to compare evaluations of `_fn` at different points because each evaluation uses a different approximation to the Fisher information matrix. 

I agree that if `log_prob_params` changes, then we would want a fresh approximation to the Fisher matrix.